variable :rabbit_delay=1,2,3

colonelfazackerley opened the pull request:
Hope I didn't make too many errors in coding style. "make test" is good (after clean).

colonelfazackerley commented:
I tested with this level
```
:num_rabbits=6
:bash=1
:rabbit_delay=1,2,10
:dig=1
:bridge=1
:climb=1
:explode=10
Q                        
                         
####################### #
                        #
# #######################
#                       O
#########################
```

andybalaam commented:
Please could you experiment with rebasing to clean up the "correcting" commits?

I suggest:

    git checkout -b my-backup-branch     # Make a branch containing the current state in case of screwing up
    git checkout master                  # Go back onto the branch we want to make a pull request from[1]
    git rebase -i origin/master          # Ask to change history

In the editor that pops up, experiment with "f" for fixup changes.

[1]Note I'd suggest normally not using master for this, but pushing another branch to GitHub to turn into a pull request.

andybalaam commented:
At this point I'd like to plug shamelessly my videos on using Git: https://www.youtube.com/watch?v=Di_-HAC6ms8&list=PLgyU3jNA6VjSUZPhZ7WtoBUnxzChBwKuw

colonelfazackerley commented:
I think I managed to tidy it. I think that if you pull
19d53fa variable :rabbit_delay=1,2,3 
that's all my work with fixup'd commits.

optionally pull
95982f3 TODO about test for double entry in level
which is the same as
9053523 Merge after fixup of multiple commits
but has a commit message that make sense.

andybalaam commented:
It looks like there are more commits on this branch now.  I was looking for fewer :-) Did you rebase -i and use the "f"?

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/Entrance.java:
```diff
@@ -31,6 +34,7 @@ public void calcNewState( World world )
     @Override
     public void step( World world )
     {
+        
```
Whitespace changes should be in separate commits (in fact, this one is probably not needed).

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/Entrance.java:
```diff
@@ -38,8 +42,14 @@ public void step( World world )
+            int delayIndex; 
+            
+            delayIndex = rabbitEntranceCount >= delay.length ? delay.length-1 : rabbitEntranceCount ;
+		    
```
This block should probably go in a separate method.

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/textworld/LineProcessor.java:
```diff
@@ -99,6 +101,21 @@ public int metaInt( String key, int def )
+        {
+            return def;
+        }
+    	int[] ret = new int[temp.size()];
```
Missing space?

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java:
```diff
@@ -103,7 +106,7 @@ public static World createWorldWithName(
             num_rabs,
             processor.metaInt( num_to_save,  1 ),
-            processor.metaInt( rabbit_delay, 4 ),
+            processor.metaIntArray( rabbit_delay, new int[4] ),
```
Haven't thought through the logic yet, but do you mean new int[] {4} here?

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java:
```diff
@@ -125,7 +128,7 @@ public static World createEmptyWorld( int width, int height )
             0,
             1,
-            4,
+            new int[4],
```
and here?

andybalaam commented on rabbit-escape-engine/src/rabbitescape/engine/textworld/TextWorldManip.java:
```diff
@@ -216,6 +219,16 @@ private static String metaLine( String name, int value )
+    	{
+    		ret = ret + "," + Integer.toString( value[i] ) ;
+    	}
+        return metaLine( name, ret); 
```
Extra space before return?

andybalaam commented:
I'd also like some tests for this code.  Might be tricky to see how/where to insert them.  If you have trouble let me know and I'll take a look.

andybalaam commented:
The comments above are a quick scan through, not me properly trying it out and thinking it through.  The idea looks sound, and the implementation strategy looks good - now I I just want to make sure it's tidy and tested.  Thanks!

colonelfazackerley commented:
I did rebase as you suggested: on my working copy. Problem is I had already pushed to my github fork. Dont know how to rebase that. Can you selectively pull certain commits? if you can, I think we are OK. If there is still a problem, I can delete everything and re-fork, and re-clone my fork.

I have some time to work on your suggestions this evening.

colonelfazackerley commented:
whitespace change at line 37 in entrance java was an accident. I declared delayIndex there, but decided reduced scope was wise.

moved block to a new private method, birthRabbit.

"Missing space before int[] ?". do you mean newline in here?
```
        }
    	int[] ret = new int[temp.size()];
```
I have not made a change, not sure I understand.

Yes it makes more sense to have a constant (nonzero!) rabbit delay for the defaults. done.

re: missing space. it looks lined up to me (using Eclipse). It's because your code uses spaces and my eclipse is putting in tabs. I will look into eclipse settings.

colonelfazackerley commented:
All your comments actioned, I think.

colonelfazackerley commented:
What I did not do the first time is "git push -f", like it says at the end of this page
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

colonelfazackerley commented:
when I try to rebase again I get something like this every time

cf@re:~/re/rabbit-escape$ git rebase -i HEAD~10
error: could not apply 691f4a7... Update download version number to 0.4.2

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 691f4a7d249a160c099f40ef951043aceebac766... Update download version number to 0.4.2

I have no idea how to resolve.

tttppp commented:
I've tried to do the rebase - the results are in this pull request: https://github.com/andybalaam/rabbit-escape/pull/42 (on my branch colonelfazackerley_rebase).  If you want to create a pull request for Andy using by fetching and re-pushing my changes then please do!

colonelfazackerley commented:
I think managed to do the merging required in the rebase and push -f back to my fork.

Thanks tttppp, but I need to figure this out myself.

andybalaam commented:
I still see this as 2 commits: 

- descriptions for all levels in the hard set
- fix typos in description key 

and I also see one of my commits in the "commits" tab above....?

andybalaam commented:
When confronted with "could not apply" try git mergetool.  I suggest installing and using KDiff3 as your merge tool.

andybalaam commented:
On white space - yes, always spaces not tabs, and always 4 spaces for each indent, please.

andybalaam commented:
Don't think I want birthRabbit to return a world, unless you have a good reason?

andybalaam commented:
When we manage to sort the Git issues, I will merge it.  I will add a test that round-trips a world containing multiple delays, probably in TestTextWorldManip, unless you want to do it...

andybalaam commented:
Try:

    git rebase -i origin/master

and put an "f" next to the commit entitled "fix typos in description key", and I think I can pull.

colonelfazackerley commented:
01. I did the 3 delay thing in my test
02. re the "desc..." and "fix typos..." commits. you already pulled these in earlier requests. i guess I can rebase more, but that will give your intermediate commits new SHA1 hashes, wont it? could you then pull the rebased repo? when i push i have to "-f" it.
03. I figured out that the rebase was resulting in a manual merge situation. I resolved it. I have also watched all of your git videos. i am now a better person.
04. re: whitespace/tabs. I fixed this in my eclipse format settings. i think you said you use eclipse? could you export your formatter settings? there are a few things I have not managed to get right.
05. i changed birthRabbit to return void. this was a thinko from octave, where nothing can be passed by reference.

colonelfazackerley commented:
just going to try this: git rebase -i origin/master

I think i have to reorder, so the "fix" commit is below the first desc commit. i think other stuff will then get new hashes. let's see...

colonelfazackerley commented:
```
cf@re:~/re/rabbit-escape$ git rebase -i HEAD~6 origin/master
[detached HEAD 2b58bb5] descriptions for all levels in the hard set
 Date: Mon Sep 14 18:28:05 2015 +0100
 20 files changed, 20 insertions(+)
Successfully rebased and updated detached HEAD.
cf@re:~/re/rabbit-escape$ git st
HEAD detached from refs/heads/master
nothing to commit, working directory clean
```

did that. now I have a detached HEAD...

andybalaam commented:
I am andybalaam on Skype, andybalaam@xmpp.org.uk on Jabber.

andybalaam commented:
I can't understand what's going on.  It looks like your rebases have never properly completed.  Let's continue the conversation on the new pull request.  Ping me if you have jabber.
